drivers/flash/flash_stm32h7x.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2020 Vossloh Cogifer
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <string.h>
+#include <drivers/flash.h>
+#include <init.h>
+#include <soc.h>
+
+#include "flash_stm32.h"
+
+#define LOG_DOMAIN flash_stm32h7
+#define LOG_LEVEL CONFIG_FLASH_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_DOMAIN);
+#define STM32H7X_SECTOR_SHIFT 17
+
+bool flash_stm32_valid_range(const struct device *dev, off_t offset,
+			     uint32_t len,
+			     bool write)
+{
+	ARG_UNUSED(write);
+
+	return flash_stm32_range_exists(dev, offset, len);
+}
+
+static int erase_sector(const struct device *dev, uint32_t sector)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	int rc;
+	uint32_t tmp;
+	volatile uint32_t *cr;
+
+	if (sector <= 7) {
+		cr = &regs->CR1;
+	} else {
+		cr = &regs->CR2;
+	}
+
+	/* if the control register is locked, do not fail silently */
+	if (*cr & FLASH_CR_LOCK) {
+		return -EIO;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+		*cr &= FLASH_CR_SNB;
+		*cr |= (FLASH_CR_SER
+			| ((sector << FLASH_CR_SNB_Pos) & FLASH_CR_SNB));
+		*cr |= FLASH_CR_START;
+		/* flush the register write */
+		tmp = *cr;
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	*cr &= ~(FLASH_CR_SER | FLASH_CR_SNB);
+
+	return rc;
+}
+
+static unsigned int get_sector(off_t offset)
+{
+	return offset >> STM32H7X_SECTOR_SHIFT;
+}
+
+int flash_stm32_block_erase_loop(const struct device *dev,
+				 unsigned int offset,
+				 unsigned int len)
+{
+	int i;
+	int rc = 0;
+
+	for (i = get_sector(offset); i <= get_sector(offset + len - 1); ++i) {
+		rc = erase_sector(dev, i);
+		if (rc < 0) {
+			break;
+		}
+	}
+	return rc;
+}
+
+int flash_stm32_check_status(const struct device *dev)
+{
+	uint32_t const error_bank1 = FLASH_FLAG_ALL_ERRORS_BANK1;
+	uint32_t const error_bank2 = FLASH_FLAG_ALL_ERRORS_BANK2;
+
+	if (FLASH_STM32_REGS(dev)->SR1 & error_bank1) {
+		LOG_DBG("Status Bank1: 0x%08x",
+			FLASH_STM32_REGS(dev)->SR1 & error_bank1);
+		return -EIO;
+	}
+
+	if (FLASH_STM32_REGS(dev)->SR2 & error_bank2) {
+		LOG_DBG("Status Bank2: 0x%08x",
+			FLASH_STM32_REGS(dev)->SR2 & error_bank2);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int wait_write_queue(const struct device *dev, off_t offset)
+{
+	int64_t timeout_time = k_uptime_get() + 100;
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	uint32_t sector = get_sector(offset);
+	volatile uint32_t *sr;
+
+	if (sector <= 7) {
+		sr = &regs->SR1;
+	} else {
+		sr = &regs->SR2;
+	}
+
+	while (*sr & FLASH_SR_QW) {
+		if (k_uptime_get() > timeout_time) {
+			LOG_ERR("Timeout! val: %d", 100);
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+static int write_4dwords(const struct device *dev,
```

Comment:
Some H7 references will use 128 bits programming length

Reply:
Rewriting the function to write_ndwords and taking an extra parameters n to define the number of double words to be written